src/workerd/io/worker-fs.h

```diff
@@ -0,0 +1,192 @@
+    void setPathForKnownRoot(KnownRoot root, kj::StringPtr path);
+    kj::Own<FsMap> finish();
+
+   private:
+    kj::Own<FsMap> map;
+  };
+
+  static kj::Own<FsMap> createDefault();
+
+ private:
+  kj::HashMap<KnownRoot, jsg::Url> paths;
+  friend class Builder;
+};
+
+// The directory delegate is used to provide the underlying implementation of
+// a directory in the WorkerFileSystem. It implements a subset of the kj::Directory
+// API focused on the specific methods that are needed for the WorkerFileSystem.
+// We currently have three DirectoryDelegate implementations:
+//   1. InMemoryDelegate: A virtual directory that is backed by an in-memory
+//      map. This is used for the interim directory structure when the FsMap
+//      if configured to use nested paths. For instance, if a known root path
+//      is mapped to be `/mybundle/modules`, then the `/mybundle` directory is
+//      backed by an InMemoryDelegate that contains the `/modules` directory.
+//   2. TempDirDelegate: A virtual directory that uses kj::InMemoryDirectory
+//      and kj::InMemoryFile to implement temporary file handling. The behavior
+//      of this directory is dependent on the current IoContext (or lack thereof).
+//   3. The BundleDirectoryDelegate (see bundle-fs.h and bundle-fs.c++) is a
+//      read-only directory implementation that is backed by the worker bundle
+//      configuration.
+//
+// The DirectoryDelegate exposes a subset of the kj::Directory API that covers
+// the functionality we support.
+class DirectoryDelegate {
+ public:
+  virtual ~DirectoryDelegate() noexcept(false) = default;
+
+  virtual kj::Own<const DirectoryDelegate> clone() const = 0;
+
+  virtual kj::FsNode::Metadata stat() const {
+    return kj::FsNode::Metadata{
+      .type = kj::FsNode::Type::DIRECTORY,
+      .size = 0,
+      .spaceUsed = 0,
+      .lastModified = kj::UNIX_EPOCH,
+      .linkCount = 1,
+      .hashCode = 0,
+    };
+  }
+
+  virtual bool exists(kj::PathPtr path) const = 0;
+
+  virtual kj::Array<kj::String> listNames() const = 0;
+
+  virtual kj::Array<kj::ReadableDirectory::Entry> listEntries() const = 0;
+
+  virtual kj::Maybe<kj::FsNode::Metadata> tryLstat(kj::PathPtr path) const = 0;
+
+  virtual kj::Maybe<kj::Own<const kj::ReadableFile>> tryOpenFile(kj::PathPtr path) const = 0;
+
+  virtual kj::Maybe<kj::Own<const kj::ReadableDirectory>> tryOpenSubdir(kj::PathPtr path) const = 0;
+
+  virtual kj::Maybe<kj::Own<const kj::File>> tryOpenFile(
+      kj::PathPtr path, kj::WriteMode mode) const {
+    return kj::none;
+  }
+
+  virtual kj::Maybe<kj::Own<kj::AppendableFile>> tryAppendFile(
+      kj::PathPtr path, kj::WriteMode mode) const {
+    return kj::none;
+  }
+
+  virtual kj::Maybe<kj::Own<const kj::Directory>> tryOpenSubdir(
+      kj::PathPtr path, kj::WriteMode mode) const {
+    return kj::none;
+  }
+
+  virtual bool tryRemove(kj::PathPtr path) const {
+    return false;
+  }
+
+  virtual kj::Own<const kj::File> createTemporary() const {
```

Comment:
```suggestion
  virtual kj::Own<const kj::File> createTemporary() {
```

Reply:
-1 ... again, modeled this directly from `kj/filesystem` to remain as consistent as possible.